### PR TITLE
Chart: Introduce value to override target namespace

### LINKED ISSUE #20511: CFP: Add a value for overriding the namespace in the helm chart.
## Cilium Feature Proposal
        
Thanks for taking time to make a feature request for Cilium! If you have usage questions, please try the [slack channel](http://slack.cilium.io/) and see the [FAQ](https://github.com/cilium/cilium/issues?utf8=%E2%9C%93&q=is:issue+label:kind/question+) first.

**Is your feature request related to a problem?**

Adding cilium to an umbrella helm chart but installing to a different namespace isn't currently possible.

**Describe the feature you'd like**

Adding `namespaceOverride: ""` to values.yaml (similar to how bitnami makes their helm charts) or how cert-manager adds the 'namespace' value.

bitnami example:
```
## @param namespaceOverride String to fully override common.names.namespace
##
namespaceOverride: ""
```

cert-manager example: 
```
# This namespace allows you to define where the services will be installed into
# if not set then they will use the namespace of the release
# This is helpful when installing cert manager as a chart dependency (sub chart)
namespace: ""
```

**(Optional) Describe your proposed solution**

I think either the way bitnami or cert-manager handles it works well and could be useful.

### PR DESCRIPTION
**Pull request by thewilli**
The previous version solely relied on the `--namespace` parameter of Helm, making it difficult to use it alongside other dependencies in an Umbrella Chart.

The new, optional `namespaceOverride` value parameter, if set, takes precedence over the Release namespace for all contained resources.

Please ensure your pull request adheres to the following guidelines:

- [x] For first time contributors, read [Submitting a pull request](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#submitting-a-pull-request)
- [x] All code is covered by unit and/or runtime tests where feasible.
- [x] All commits contain a well written commit description including a title,
      description and a `Fixes: #XXX` line if the commit addresses a particular
      GitHub issue.
- [x] If your commit description contains a `Fixes: <commit-id>` tag, then
      please add the commit author[s] as reviewer[s] to this issue.
- [x] All commits are signed off. See the section [Developer’s Certificate of Origin](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#dev-coo)
- [x] Provide a title or release-note blurb suitable for the release notes.
- [x] Are you a user of Cilium? Please add yourself to the [Users doc](https://github.com/cilium/cilium/blob/main/USERS.md)
- [x] Thanks for contributing!

Fixes #20511

```release-note
Added Helm Chart value for overriding target namespace.
```

### CODE REVIEW

File: USERS.md
Comment (sayboras) [inline USERS.md]: can you split this change into a separate PR, so that it will be easier and faster to review and merge ? Thanks

**Review by sayboras — COMMENTED** [review]
I have one comment as per below.

The changes look reasonable for envoy related files. Thanks.

**Comment by sayboras** [comment]
Also, can you run the below to fix the chart linting issue?

https://github.com/cilium/cilium/actions/runs/10637360015/job/29504666086?pr=34624
```
 1 chart(s) linted, 0 chart(s) failed
make: Leaving directory '/home/runner/work/cilium/cilium/install/kubernetes'
please run 'make -C install/kubernetes' and submit your changes
```

**Comment by thewilli** [comment]
> Also, can you run the below to fix the chart linting issue?
> 
> https://github.com/cilium/cilium/actions/runs/10637360015/job/29504666086?pr=34624
> 
> ```
>  1 chart(s) linted, 0 chart(s) failed
> make: Leaving directory '/home/runner/work/cilium/cilium/install/kubernetes'
> please run 'make -C install/kubernetes' and submit your changes
> ```

@sayboras  done

File: USERS.md
Comment (thewilli) [reply USERS.md]: done

**Review by sayboras — APPROVED** [review]
Looks good to me for envoy related files. Thanks.

**Review by ghost — APPROVED** [review]
Thanks for the PR @thewilli, looks good to me.

**Review by rolinh — APPROVED** [review]
lgtm, thanks!

File: install/kubernetes/cilium/templates/_helpers.tpl
Comment (giorio94) [inline install/kubernetes/cilium/templates/_helpers.tpl]: Nit. Could be shortened to:

```suggestion
{{- .Values.namespaceOverride | default .Release.Namespace -}}
```

File: install/kubernetes/cilium/templates/cilium-agent/role.yaml
Comment (giorio94) [inline install/kubernetes/cilium/templates/cilium-agent/role.yaml]: Is it there any specific reason for using `template` rather than `include`? `include` would not require to reason about whether the output is going to be used as part of a pipeline or not.

File: install/kubernetes/cilium/values.yaml.tmpl
Comment (giorio94) [inline install/kubernetes/cilium/values.yaml.tmpl]: Is this change intentional?

**Review by giorio94 — APPROVED** [review]
Thanks, /lgtm, I've left a couple of minor nits inline.

I wonder whether it could make sense to have a basic linter as part of CI to verify that we don't accidentally use `.Release.Namespace` instead of the dedicated template. Maybe something as basic as checking the output of `grep -R --exclude=... '.Release.Namespace' install/kubernetes` with appropriate exclusions.

File: install/kubernetes/cilium/values.yaml.tmpl
Comment (thewilli) [reply install/kubernetes/cilium/values.yaml.tmpl]: I did not explicitly change this part, so I assume it resulted in one of the generative processes

Comment (giorio94) [reply install/kubernetes/cilium/values.yaml.tmpl]: Hmm, this file is actually the source of truth for a few other autogenerated ones, but maybe it happened during conflict resolution. Could you please revert this change and rerun the generation then?

**Review by joamaki — APPROVED** [review]

**Review by youngnick — APPROVED** [review]
LGTM for servicemesh files.

**Comment by giorio94** [comment]
I've dropped the review-request for the Community team, as the `USERS.md` change has been removed from this PR.

**Comment by thewilli** [comment]
@squeed any chance to give this a quick review? thank you 🙂

**Comment by thewilli** [comment]
@giorio94 @sayboras  Do we have any chance of getting this merged, so it can become part of the next release? It would really help to progress with my stack, and the whole PR is only waiting for a single approval for 3 weeks now.

**Comment by giorio94** [comment]
@squeed, Gentle ping :pray: 

@thewilli Could you please rebase this PR onto main? A few of the Ginkgo tests are failing pretty consistently, and a rebase would most likely fix them.

**Pull request merged**
